README.md
# SniperTips

[![PRs Welcome](https://img.shields.io/badge/PRs-welcome-brightgreen.svg?style=flat-square)](http://makeapullrequest.com)
![GitHub tag (latest SemVer)](https://img.shields.io/github/tag/ps-wow/SniperTips.svg)

A tooltip framework for adding information onto tooltips.

## Addons using this framework

 - [SniperTips_HunterPetFood](https://github.com/ps-wow/SniperTips_HunterPetFood)
 - [SniperTips_VendorPrice](https://github.com/ps-wow/SniperTips_VendorPrice)


CHANGELOG.md
# SniperTips Changes

All notable changes to this project will be documented in this file.

The format is based on [Keep a Changelog](https://keepachangelog.com/en/1.0.0/),
and this project adheres to [Semantic Versioning](https://semver.org/spec/v2.0.0.html).

## [v2.0.2] - 2019-08-10
### Fixed
 - Additional fix for previous #4 fix, specifically for debuffs

## [v2.0.1] - 2019-08-09
### Fixed
 - Fix issue #4 - error when hovering over buff or debuff

## [v2.0.0] - 2019-07-14
### Added
 - Add ace configuration
 - Add handling of spell and unit tooltips

### Changed
 - Change way item handlers are registered, instead of passing lots of params pass just self and item/spell/unit


## [v1.0.3] - 2019-06-24
### Changed
 - Ace3 dependency moved to optional.

## [v1.0.2] - 2019-06-23
### Changed
 - Update toc to match wow 1.13.2
 - Embed LibStub

## [v1.0.1] - 2019-06-02
### Fixed
 - [Issue 3](https://github.com/ps-wow/SniperTips/issues/3) - Bug: Clicking on an achievement throws an error message [retail/8.1.5]

## [v1.0.0] - 2019-05-28
### Added
 - Add ability for addons to register as item handlers, these methods will receive params from GetItemInfo()

main.lua
-- Library Version
local MAJOR, MINOR = 1, 0
local SniperTips = LibStub:NewLibrary("SniperTips-1.0", MAJOR, MINOR);

SniperTips.kbDEBUG = true

if not SniperTips then
  return -- already loaded and no upgrade necessary
end

function SniperTips:Dump(str, obj)
  if ViragDevTool_AddData and SniperTips.kbDEBUG then
      ViragDevTool_AddData(obj, str)
  end
end

SniperTips.handlers = SniperTips.handlers or {}
SniperTips.handlers.items = SniperTips.handlers.items or {}

if not SniperTips.frame then
  SniperTips.frame=CreateFrame("Frame", SniperTips)
end

local tooltipTypes = {
  "GameTooltip",
  "ItemRefTooltip",
};

-- Options without TipTac. If the base TipTac addon is used, the global TipTac_Config table is used instead
local cfg = {
  if_enable = true,
  if_infoColor = { 0.2, 0.6, 1 },
  if_itemQualityBorder = true,

  if_showCurrencyId = true,					-- Az: no option for this added to TipTac/options yet!
  if_showAchievementIdAndCategory = false,	-- Az: no option for this added to TipTac/options yet!
  if_showIcon = true,
  if_smartIcons = true,
  if_borderlessIcons = false,
  if_iconSize = 42,
};

local tipsToAddIcon = tooltipTypes;
local tipDataAdded = {}; -- Sometimes, OnTooltipSetItem/Spell is called before the tip has been filled using SetHyperlink, we use the array to test if the tooltip has had data added

function SniperTips:GetVersion()
  return MAJOR, MINOR
end

function SniperTips:AddItemHandler(Addon)
  table.insert(SniperTips.handlers.items, Addon.name)
end

-- OnTooltipSetItem
local function OnTooltipSetItem(self,...)
  if (cfg.if_enable) and (not tipDataAdded[self]) then
    local _, link = self:GetItem();
    if (link) then
      local linkType, id = link:match("H?(%a+):(%d+)");
      if (id) then
        tipDataAdded[self] = linkType;
        SniperTips:HandleItem(self,link,linkType,id);
      end
    end
  end
end

local function OnTooltipCleared(self)
  tipDataAdded[self] = nil;
  if (self.SetIconTextureAndText) then
    self:SetIconTextureAndText();
  end
end

-- Resolve the tooltips from _G
local function ResolveGlobalNamedObjects(tipTable)
  local resolved = {};
  for index, tipName in ipairs(tipTable) do
    -- lookup the global object from this name, assign false if nonexistent, to preserve the table entry
    local tip = (_G[tipName] or false);

    -- Check if this object has already been resolved. This can happen for thing like AtlasLoot, which sets AtlasLootTooltip = GameTooltip
    if (resolved[tip]) then
      tip = false;
    elseif (tip) then
      resolved[tip] = index;
    end

    -- Assign the resolved object or false back into the table array
    tipTable[index] = tip;
  end
end

-- HOOK: ItemRefTooltip + GameTooltip: SetHyperlink
local function SetHyperlink_Hook(self,hyperLink)
  if (cfg.if_enable) and (not tipDataAdded[self]) then
    local refString = hyperLink:match("|H([^|]+)|h") or hyperLink;
    local linkType = refString:match("^[^:]+");

    SniperTips:Dump('linkType', linkType)

    -- Call Relevant handler

    -- Only add to the tooltip if it has lines?
    if (self:NumLines() > 0) then
      if (linkType == 'item') then
        tipDataAdded[self] = "hyperlink";
        SniperTips:HandleItem(self,refString,(":"):split(refString));
      elseif (linkType == 'achievement') then
      else
      end
    end
  end
end

-- Frame Management

function SniperTips.frame:DoHooks()
  for index, tip in ipairs(tooltipTypes) do
    if (type(tip) == "table") and (type(tip.GetObjectType) == "function") and (tip:GetObjectType() == "GameTooltip") then
      -- if (tipsToAddIcon[tip:GetName()]) then
      --   self:CreateTooltipIcon(tip);
      -- end
      hooksecurefunc(tip,"SetHyperlink",SetHyperlink_Hook);
      tip:HookScript("OnTooltipSetItem",OnTooltipSetItem);
      tip:HookScript("OnTooltipCleared",OnTooltipCleared);
    end
  end
end

function SniperTips.frame:OnApplyConfig()
  local gameFont = GameFontNormal:GetFont();
  for index, tip in ipairs(tooltipTypes) do
    if (type(tip) == "table") and (tipsToAddIcon[tip:GetName()]) and (tip.ttIcon) then
      if (cfg.if_showIcon) then
        tip.ttIcon:SetSize(cfg.if_iconSize,cfg.if_iconSize);
        tip.ttCount:SetFont(gameFont,(cfg.if_iconSize / 3),"OUTLINE");
        tip.SetIconTextureAndText = SetIconTextureAndText;
        if (cfg.if_borderlessIcons) then
          tip.ttIcon:SetTexCoord(0.07,0.93,0.07,0.93);
        else
          tip.ttIcon:SetTexCoord(0,1,0,1);
        end
      elseif (tip.SetIconTextureAndText) then
        tip.ttIcon:Hide();
        tip.SetIconTextureAndText = nil;
      end
    end
  end
end

SniperTips.frame:SetScript("OnEvent",function(self,event,...)
  -- What tipsToModify to use, TipTac's main addon, or our own?
  if (TipTac and TipTac.tipsToModify) then
    tooltipTypes = TipTac.tipsToModify;
  else
    ResolveGlobalNamedObjects(tooltipTypes)
  end

  -- Use TipTac settings if installed
  if (TipTac_Config) then
    cfg = TipTac_Config;
  end

  -- Hook tips and apply settings
  SniperTips.frame:DoHooks();
  SniperTips.frame:OnApplyConfig();

  -- Cleanup; we no longer need to receive any events
  SniperTips.frame:UnregisterAllEvents();
  SniperTips.frame:SetScript("OnEvent",nil);
end);

SniperTips.frame:SetScript("OnUpdate", OnUpdate);

SniperTips.frame:RegisterEvent("VARIABLES_LOADED");

-- Other Addon Stuff

function SniperTips:HandleItem(self,link,linkType,id)
  local itemName, itemLink, itemRarity, itemLevel, itemMinLevel, itemType, itemSubType, itemStackCount,
  itemEquipLoc, itemIcon, itemSellPrice, itemClassID, itemSubClassID, bindType, expacID, itemSetID,
  isCraftingReagent = GetItemInfo(link);

  for _, addonName in ipairs(SniperTips.handlers.items) do
    a = LibStub("AceAddon-3.0"):GetAddon(addonName);
    a:HandleItem(self, itemName, itemLink, itemRarity, itemLevel, itemMinLevel, itemType, itemSubType, itemStackCount,
    itemEquipLoc, itemIcon, itemSellPrice, itemClassID, itemSubClassID, bindType, expacID, itemSetID, 
    isCraftingReagent)
    --handler(GetItemInfo(link))
  end
end


addon.lua
-- SniperTips Addon
--local SniperTipsAddon = LibStub()
local SniperTips_Core = LibStub("AceAddon-3.0"):NewAddon("SniperTips_Core")
local LibTooltip = LibStub("SniperTips-2.0");

SniperTips_Core.debug = true

function SniperTips_Core:Dump(str, obj)
  if ViragDevTool_AddData and SniperTips_Core.debug then
      ViragDevTool_AddData(obj, str)
  end
end

function SniperTips_Core:GetTipColour(cfg)
  return cfg.tipColour[1], cfg.tipColour[2], cfg.tipColour[3]
end

function SniperTips_Core:HandleItem(self, item, config)
  if (self.AddDoubleLine ~= nil) then
    if (config.showIds == true) then
      local r, g, b = SniperTips_Core:GetTipColour(config)

      self:AddDoubleLine('Item ID:', item.id, r, g, b, r, g, b)
    end
  end
end

function SniperTips_Core:HandleSpell(self, spell, config)
  if (self.AddDoubleLine ~= nil) then
    if (config.showIds == true) then
      local r, g, b = SniperTips_Core:GetTipColour(config)

      self:AddDoubleLine('Spell ID:', spell.id, r,g,b, r,g,b)
    end
  end
end

function SniperTips_Core:HandleUnit(self, unit, config)
  if (self.AddDoubleLine ~= nil) then
    if (config.showIds == true) then
      local r, g, b = SniperTips_Core:GetTipColour(config)
      self:AddDoubleLine('Unit ID:', 'TODO', r,g,b, r,g,b)
    end
  end
end

function SniperTips_Core:HandleBuff(self, buff, config)
  if (self.AddDoubleLine ~= nil) then
    if (config.showIds == true) then
      local r,g,b = SniperTips_Core:GetTipColour(config)
      SniperTips_Core:Dump('spellid', buff.spell.id)
      self:AddLine('Spell ID: ' .. buff.spell.id, r,g,b)
    end
  end
end

function SniperTips_Core:HandleDebuff(self, debuff, config)
  if (self.AddDoubleLine ~= nil) then
    if (config.showIds == true) then
      local r,g,b = SniperTips_Core:GetTipColour(config)
      self:AddLine('Spell ID: '.. debuff.spell.id, r,g,b)
    end
  end
end

LibTooltip:AddItemHandler(SniperTips_Core)
LibTooltip:AddSpellHandler(SniperTips_Core)
LibTooltip:AddUnitHandler(SniperTips_Core)
LibTooltip:AddBuffHandler(SniperTips_Core)
LibTooltip:AddDebuffHandler(SniperTips_Core)


config.lua
-- SniperTips Configuration
local SniperTipsConfig = LibStub("AceAddon-3.0"):NewAddon("SniperTips_Config")
local defaults = {
  profile = {
    enabled = true,
    showIds = true,
    tipColour = { r = 0.2, g = 0.6, b = 1 }
  }
}

function SniperTipsConfig:OnInitialize()
  -- Register AceDB with defaults
  self.db = LibStub("AceDB-3.0"):New("SniperTips_DB", defaults, true)

  -- AceDBOptions
  local options = {
    type = "group",
    args = {
      enabled = {
        name = "Enabled",
        desc = "Enables / disables the addon",
        type = "toggle",
        handler = SniperTipsConfig,
        get = 'GetEnabled',
        set = 'SetEnabled'
      },
      showIds = {
        name = "Show IDs",
        desc = "Shows Spell and Item IDs",
        type = "toggle",
        handler = SniperTipsConfig,
        get = 'GetShowIDs',
        set = 'SetShowIDs'
      },
      tipColour = {
        name = "Default Tooltip Colour",
        desc = "Set the default tooltip colour",
        type = "color",
        handler = SniperTipsConfig,
        get = 'GetTooltipColour',
        set = 'SetTooltipColour'
      }
    }
  }

  options.args.profiles = LibStub("AceDBOptions-3.0"):GetOptionsTable(self.db)

  --
  -- ACE CONFIG
  --
  local config = LibStub("AceConfig-3.0")
  config:RegisterOptionsTable("SniperTips", options, {"/snipertips", "/st"})

  -- AceConfigDialog
  LibStub("AceConfigDialog-3.0"):AddToBlizOptions("SniperTips")
end

function SniperTipsConfig:GetDB()
  return self.db
end

function SniperTipsConfig:GetProfile()
  local profile = self.db:GetCurrentProfile()
  local config = self.db.profiles[profile]

  if (not config) then
    config = {}
  end

  for k,v in pairs(defaults.profile) do
    if (not config[k]) then
      config[k] = v
    end
  end
  return config
end

-- ACE DB OPTIONS

-- Enabled
function SniperTipsConfig:GetEnabled(info)
  return self.db.profile.enabled
end
function SniperTipsConfig:SetEnabled(info, value)
  self.db.profile.enabled = value
end

-- ShowIDs
function SniperTipsConfig:GetShowIDs(info)
  return self.db.profile.showIds
end
function SniperTipsConfig:SetShowIDs(info, value)
  self.db.profile.showIds = value
end

-- Tooltip Colour
function SniperTipsConfig:GetTooltipColour(info)
  return unpack(self.db.profile.tipColour)
end
function SniperTipsConfig:SetTooltipColour(info, r, g, b)
  self.db.profile.tipColour = { r, g, b }
end

_G['SniperTips_Config'] = SniperTipsConfig


lib.lua
-- SniperTips Library

-- Library Version
local MAJOR, MINOR = 1, 0
local SniperTips = LibStub:NewLibrary("SniperTips-2.0", MAJOR, MINOR);

SniperTips.kbDEBUG = true

if not SniperTips then
  return -- already loaded and no upgrade necessary
end

function SniperTips:Dump(str, obj)
  if ViragDevTool_AddData and SniperTips.kbDEBUG then
      ViragDevTool_AddData(obj, str)
  end
end

SniperTips.handlers = SniperTips.handlers or {}
SniperTips.handlers.items = SniperTips.handlers.items or {}
SniperTips.handlers.spells = SniperTips.handlers.spells or {}
SniperTips.handlers.units = SniperTips.handlers.units or {}
SniperTips.handlers.buffs = SniperTips.handlers.buffs or {}
SniperTips.handlers.debuffs = SniperTips.handlers.debuffs or {}

if not SniperTips.frame then
  SniperTips.frame=CreateFrame("Frame", SniperTips)
end

local tooltipTypes = {
  "GameTooltip",
  "ItemRefTooltip",
};

-- Options without TipTac. If the base TipTac addon is used, the global TipTac_Config table is used instead
local cfg = {
  if_enable = true,
  if_infoColor = { 0.2, 0.6, 1 },
  if_itemQualityBorder = true,

  if_showCurrencyId = true,					-- Az: no option for this added to TipTac/options yet!
  if_showAchievementIdAndCategory = false,	-- Az: no option for this added to TipTac/options yet!
  if_showIcon = true,
  if_smartIcons = true,
  if_borderlessIcons = false,
  if_iconSize = 42,
};

local tipsToAddIcon = tooltipTypes;
local tipDataAdded = {}; -- Sometimes, OnTooltipSetItem/Spell is called before the tip has been filled using SetHyperlink, we use the array to test if the tooltip has had data added

function SniperTips:GetVersion()
  return MAJOR, MINOR
end

function SniperTips:AddItemHandler(Addon)
  table.insert(SniperTips.handlers.items, Addon.name)
end

function SniperTips:AddSpellHandler(Addon)
  table.insert(SniperTips.handlers.spells, Addon.name)
end

function SniperTips:AddUnitHandler(Addon)
  table.insert(SniperTips.handlers.units, Addon.name)
end

function SniperTips:AddBuffHandler(Addon)
  table.insert(SniperTips.handlers.buffs, Addon.name)
end

function SniperTips:AddDebuffHandler(Addon)
  table.insert(SniperTips.handlers.debuffs, Addon.name)
end

-- OnTooltipSetItem
local function OnTooltipSetItem(self,...)
  if (not tipDataAdded[self]) then
    local _, link = self:GetItem();
    if (link) then
      local linkType, id = link:match("H?(%a+):(%d+)");
      if (id) then
        tipDataAdded[self] = linkType;
        SniperTips:HandleItem(self,link,linkType,id);
      end
    end
  end
end

local function OnTooltipSetSpell(self,...)
  if (not tipDataAdded[self]) then
    local spell, id = self:GetSpell();
    SniperTips:HandleSpell(self,spell,id)
  end
end

local function OnTooltipSetUnit(self,...)
  if (not tipDataAdded[self]) then
    local name, unit = self:GetUnit();
    SniperTips:HandleUnit(self,name,unit)
  end
end

local function SetUnitAura(self,unit,index,filter)
  if (not tipDataAdded[self]) then
    if (filter == 'HELPFUL') then
      SniperTips:HandleBuff(self, unit, index, filter)
    elseif (filter == 'PASSIVE') then
      SniperTips:Dump('handling passive', true)
    else
      SniperTips:Dump('filter', filter)
      SniperTips:HandleDebuff(self, unit, index, filter)
    end
  end
end

local function OnTooltipCleared(self)
  tipDataAdded[self] = nil;
  if (self.SetIconTextureAndText) then
    self:SetIconTextureAndText();
  end
end

-- Resolve the tooltips from _G
local function ResolveGlobalNamedObjects(tipTable)
  local resolved = {};
  for index, tipName in ipairs(tipTable) do
    -- lookup the global object from this name, assign false if nonexistent, to preserve the table entry
    local tip = (_G[tipName] or false);

    -- Check if this object has already been resolved. This can happen for thing like AtlasLoot, which sets AtlasLootTooltip = GameTooltip
    if (resolved[tip]) then
      tip = false;
    elseif (tip) then
      resolved[tip] = index;
    end

    -- Assign the resolved object or false back into the table array
    tipTable[index] = tip;
  end
end

-- HOOK: ItemRefTooltip + GameTooltip: SetHyperlink
local function SetHyperlink_Hook(self,hyperLink)
  if (not tipDataAdded[self]) then
    local refString = hyperLink:match("|H([^|]+)|h") or hyperLink;
    local linkType = refString:match("^[^:]+");

    -- Call Relevant handler

    -- Only add to the tooltip if it has lines?
    if (self:NumLines() > 0) then
      if (linkType == 'item') then
        tipDataAdded[self] = "hyperlink";
        SniperTips:HandleItem(self,refString,(":"):split(refString));
      elseif (linkType == 'achievement') then
      else
      end
    end
  end
end

-- Frame Management

function SniperTips.frame:DoHooks()
  for index, tip in ipairs(tooltipTypes) do
    if (type(tip) == "table") and (type(tip.GetObjectType) == "function") and (tip:GetObjectType() == "GameTooltip") then
      hooksecurefunc(tip,"SetHyperlink",SetHyperlink_Hook);
      tip:HookScript("OnTooltipSetItem",OnTooltipSetItem);
      tip:HookScript("OnTooltipSetUnit",OnTooltipSetUnit);
      tip:HookScript("OnTooltipSetSpell",OnTooltipSetSpell);
      hooksecurefunc(tip, "SetUnitAura", SetUnitAura);
      tip:HookScript("OnTooltipCleared",OnTooltipCleared);
    end
  end
end

function SniperTips.frame:OnApplyConfig()
  local gameFont = GameFontNormal:GetFont();
  for index, tip in ipairs(tooltipTypes) do
    if (type(tip) == "table") and (tipsToAddIcon[tip:GetName()]) and (tip.ttIcon) then
      if (cfg.if_showIcon) then
        tip.ttIcon:SetSize(cfg.if_iconSize,cfg.if_iconSize);
        tip.ttCount:SetFont(gameFont,(cfg.if_iconSize / 3),"OUTLINE");
        tip.SetIconTextureAndText = SetIconTextureAndText;
        if (cfg.if_borderlessIcons) then
          tip.ttIcon:SetTexCoord(0.07,0.93,0.07,0.93);
        else
          tip.ttIcon:SetTexCoord(0,1,0,1);
        end
      elseif (tip.SetIconTextureAndText) then
        tip.ttIcon:Hide();
        tip.SetIconTextureAndText = nil;
      end
    end
  end
end

SniperTips.frame:SetScript("OnEvent",function(self,event,...)
  -- What tipsToModify to use, TipTac's main addon, or our own?
  if (TipTac and TipTac.tipsToModify) then
    tooltipTypes = TipTac.tipsToModify;
  else
    ResolveGlobalNamedObjects(tooltipTypes)
  end

  -- Use TipTac settings if installed
  if (TipTac_Config) then
    cfg = TipTac_Config;
  end

  -- Hook tips and apply settings
  SniperTips.frame:DoHooks();
  SniperTips.frame:OnApplyConfig();

  -- Cleanup; we no longer need to receive any events
  SniperTips.frame:UnregisterAllEvents();
  SniperTips.frame:SetScript("OnEvent",nil);
end);

SniperTips.frame:SetScript("OnUpdate", OnUpdate);

SniperTips.frame:RegisterEvent("VARIABLES_LOADED");

-- Other Addon Stuff

function SniperTips:GetProfile()
  local config = _G['SniperTips_Config']
  return config:GetProfile()
end

-- Tooltip Handlers

function SniperTips:HandleItem(self, link, linkType, id)
  local itemName, itemLink, itemRarity, itemLevel, itemMinLevel, itemType, itemSubType, itemStackCount,
  itemEquipLoc, itemIcon, itemSellPrice, itemClassID, itemSubClassID, bindType, expacID, itemSetID,
  isCraftingReagent = GetItemInfo(link);

  local item = {
    id = id,
    name = itemName,
    link = itemLink,
    rarity = itemRarity,
    ilvl = itemLevel,
    minLevel = itemMinLevel,
    stackCount = itemStackCount,
    type = itemType,
    subType = itemSubType,
    equipLoc = itemEquipLoc,
    icon = itemIcon,
    sellPrice = itemSellPrice,
    classID = itemClassID,
    subClassID = itemSubClassID,
    bindType = bindType,
    expacID = expacID,
    setId = itemSetID,
    reagent = isCraftingReagent
  }

  local config = SniperTips:GetProfile()

  for _, addonName in ipairs(SniperTips.handlers.items) do
    local a = LibStub("AceAddon-3.0"):GetAddon(addonName);
    a:HandleItem(self, item, config)
  end
end

function SniperTips:HandleSpell(self, spell, id)
  local spellData = {
    spell = spell,
    id = id
  }

  local config = SniperTips:GetProfile()

  for _, addonName in ipairs(SniperTips.handlers.spells) do
    local a = LibStub("AceAddon-3.0"):GetAddon(addonName);
    a:HandleSpell(self, spellData, config)
  end
end

function SniperTips:HandleUnit(self, unit, name, guid)
  local config = SniperTips:GetProfile()
  for _, addonName in ipairs(SniperTips.handlers.units) do
    local a = LibStub("AceAddon-3.0"):GetAddon(addonName);
    a:HandleUnit(self, unit, config)
  end
end

function SniperTips:IsClassic()
  if (_G['GetPetHappiness'] ~= nil) then
    return true
  end
  return false
end

function SniperTips:ParseAura(unit, index, filter)
  local isClassic = SniperTips:IsClassic()
  local aura = { UnitAura(unit, index, filter) }
  local buff = {
    unit = unit,
    index = index,
    spell = {}
  }

  buff.name = aura[1]

  if (isClassic) then
    buff.type = aura[4]
    buff.duration = aura[5]
    buff.expirationTime = aura[6]
    buff.source = aura[7]
    buff.spell.id = aura[10]
    
  else
    buff.icon = aura[2]
    buff.count = aura[3]
    buff.type = aura[4]
    buff.duration = aura[5]
    buff.expirationTime = aura[6]
    buff.source = aura[7]
    buff.isStealable = aura[8]
    buff.nameplateShowPersonal = aura[9]
    buff.spell.id = aura[10]
    buff.canApplyAura = aura[11]
    buff.isBossDebuff = aura[12]
    buff.castByPlayer = aura[13]
    buff.nameplateShowAll = aura[14]
    buff.timeMod = aura[15]
  end

  return buff
end

function SniperTips:HandleBuff(self, unit, index, filter)
  local config = SniperTips:GetProfile()
  local buff = SniperTips:ParseAura(unit, index, filter)

  for _, addonName in ipairs(SniperTips.handlers.buffs) do
    local a = LibStub("AceAddon-3.0"):GetAddon(addonName);
    a:HandleBuff(self, buff, config)
  end
end

function SniperTips:HandleDebuff(self, unit, index, filter)
  local config = SniperTips:GetProfile()
  local debuff = SniperTips:ParseAura(unit, index, filter)

  for _, addonName in ipairs(SniperTips.handlers.debuffs) do
    local a = LibStub("AceAddon-3.0"):GetAddon(addonName);
    a:HandleDebuff(self, debuff, config)
  end
end
